---
title: "RHAPSODY WP3 Pre-Diabetes"
subtitle: "How-To"
author: 
- name: "Mickaël Canouil, Ph.D."
  affiliation: "Université de Lille, CNRS, Institut Pasteur de Lille, UMR 1283 - EGID, F-59000, Lille, France."
  email: "mickael.canouil@cnrs.fr"
date: '`r format(Sys.time(), "%B %d, %Y")`'
output: 
  html_document:
    keep_md: false
    theme: simplex
    toc: false
    toc_depth: 4
    toc_float: 
      collapsed: false
    fig_width: 5
    fig_height: 4
    number_sections: true
    self_contained: true
    mathjax: default
    df_print: kable
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  cache = FALSE,
  echo = TRUE,
  warning = FALSE,
  error = FALSE,
  message = FALSE,
  include = TRUE,
  tidy = FALSE,
  crop = TRUE,
  autodep = TRUE,
  fig.align = "center",
  fig.pos = "!H",
  dpi = 120,
  size = "small", 
  results = "asis",
  dev = "png",
  fig.path = paste0(tempdir(), "/")
)

library(tidyverse)
library(data.tree)
library(kableExtra)
```

```{r logo, echo = FALSE}
htmltools::img(
  src = "RHAPSODY_Logo_WEB_Color.png", 
  alt = "logo", 
  style = "position:absolute; top:0; right:0; padding:10px;"
)
```


# How-to {.tabset}

**Note**: `~` in the following is `/home/rhapsody`

## Without Docker

### Files & Scripts

#### Retrieve the latest version

In order to retrieve the latest version of the main script `RHAPSODY_WP3_PreDiab.Rmd` and `utils/*` scripts, you can copy/paste the git commands below.  
This will clone (*i.e.*, download) the scripts from **GitHub**.  
Scripts will be downloaded into your home directory within a directory named `WP3`.

```{sh, eval = TRUE, echo = FALSE, results = "hide"}
rm -Rf ~/WP3
```

```{sh, eval = TRUE, results = "hide"}
git clone https://github.com/mcanouil/RHAPSODY.git ~/WP3/scripts
```

Once donwload is complete, you should get the following directory tree in your home (`` `cd ~` ``).

```{r, echo = FALSE, results = "markup"}
outout_tree_structure <- c(
  list.files(
    path = "~/WP3/scripts", 
    full.names = TRUE, include.dirs = TRUE
  ), 
  list.files(
    path = "~/WP3/scripts", 
    full.names = TRUE, include.dirs = TRUE
  ) %>%
    list.files(full.names = TRUE)
) %>%
  unlist() %>%
  data.frame(pathString = .) %>%
  as.Node()

cat(paste(sapply(capture.output(outout_tree_structure)[-1], paste0, "\n"), collapse = ""))
```

#### Description

* `~/WP3/scripts/README.pdf`  
  a short README file;
* `~/WP3/scripts/RHAPSODY_WP3_PreDiab.Rmd`  
  a Rmarkdown script which perform:
    * opal node access,
    * phenotype QC, 
    * preliminary modelling,
    * VCFs formatting/filtering,
    * variants analysis;
* `~/WP3/scripts/opal_credentials.txt`  
  the server, login and password informations to access phenotype data on your local node;
* `~/WP3/scripts/utils/RHAPSODY_WP3_PreDiab_DEBUG.R`,  
  a small part of the main Rmarkdown script, which includes only node access and phenotype QC;
* `~/WP3/scripts/utils/RunAnalysis.R`  
  a example of a R script to run the whole analysis in bash (parameters have to be filled properly according to your informations)  
  using `` `Rscript ~/WP3/scripts/utils/RunAnalysis.R` ``;
* `~/WP3/scripts/utils/handleVCF.sh`  
  the bash/shell script used to format/filter VCFs files prior to the variant analysis.  
  This bash/shell script is also included within the main script, but you can use it directly as a standalone, if so (or if your already done formatting and don't want to do it again), parameter `format_vcfs` must be set to `FALSE`).
* `~/WP3/scripts/utils/Install_Rpackages.R`  
  a R script which install a predefined version of all R packages loaded in the Rmarkdown script, only if needed (also avaiable within the main script);
* `~/WP3/scripts/utils/check_analysis.Rmd`  
  a Rmarkdown script to ''quickly'' check if the analyses went well (*i.e.*, compute quantile-quatile plot, manhattan plot and a plot to compare chromosomal position to sequencial position);
* `~/WP3/scripts/utils/README.Rmd`  
  a Rmarkdown script to create `README.pdf`.
* `~/WP3/scripts/utils/howto.Rmd`  
  a how-to document and its html version `howto.html`.

### Check environment settings

#### R and R packages version

Before runing anything, you can check the current settings of your cluster/grid/computer/laptop.  
**For reproducibility:**

* The default R version is **`r R.version.string`**.
* The default R packages are listed below.

```{r, echo = FALSE}
x <- utils::read.table(stringsAsFactors = FALSE, header = TRUE, text =
  "package version
  broom   0.5.0
  scales   1.0.0
  cowplot   0.9.3
  viridis   0.5.1
  data.tree   0.7.8
  devtools  1.13.6
  grid   3.4.2
  Hmisc   4.1-1
  kableExtra   0.9.0
  knitr    1.20
  lme4  1.1-18-1
  lmerTest   3.0-1
  parallel   3.4.2
  readxl   1.1.0
  writexl     01.0
  tidyverse   1.2.1"
)
x %>% 
  group_by(fac = cut(x = seq_len(nrow(x)), 4)) %>% 
  nest() %>% 
  .[["data"]] %>% 
  .[order(map_int(., nrow), decreasing = TRUE)] %>% 
  map(~as.data.frame(t(.x), stringsAsFactors = FALSE)) %>% 
  bind_rows() %>% 
  t() %>% 
  `colnames<-`(rep(colnames(x), 4)) %>% 
  replace_na("") %>% 
  knitr::kable(format = "html", row.names = FALSE) %>% 
  kableExtra::kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    position = "center"
  ) %>% 
  kableExtra::column_spec(column = seq(from = 1, to = 2*4, by = 2), bold = TRUE)
```

* To quickly check and install all R packages which are not installed or in a different version, you can source/run the script `~/WP3/scripts/utils/Install_Rpackages.R`.
    * Interactively using R
        ```{r, eval = FALSE}
        source("~/WP3/scripts/utils/Install_Rpackages.R")
        ```
    * Non-interactively using bash/shell
        ```{sh, eval = FALSE}
        Rscript ~/WP3/scripts/utils/Install_Rpackages.R
        ```

* To only check R packages version, you can copy/paste the following code (*i.e.*, line `#14` to line `#153` of `~/WP3/scripts/utils/Install_Rpackages.R`) and display the `data.frame` (*i.e.*, `show(list_packages)`).  
  Either you get a zero line `data.frame` (*i.e.*, everything is as it should be), or you get a line for each discrepancy (*i.e.*, package name, recommended version and your local version).
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/Install_Rpackages.R")[14:153]}
    ```


#### VCFtools version

If VCFtools is not available on your machine, please install it following the instructions on [https://vcftools.github.io/](https://vcftools.github.io/).  
**For reproducibility:**

* The default version is **`r system(command = "vcftools --version", intern = TRUE)`**.


### Run the analysis

#### OPAL database credentials

To avoid OPAL database credentials to be hard written within the script, you should create a text file with the `server`, `username` and `password` to access your local OPAL database (*i.e.*, RHAPSODY node hosting the phenotype data).  
You can also modify the default file `~/WP3/scripts/opal_credentials.txt`.

```{r, eval = FALSE, code = readLines("~/WP3/scripts/opal_credentials.txt")}
```

**NOTE:** The `username` provided in that file should have **"download"** rights.  
The data are downloaded locally in the R session for and during the analysis, then all local data are deleted (*i.e.*, the data are not leaving the R session in any way).  

#### Imputation quality (VCF)

Please check where the imputation quality is stored in your VCF files.  
Depending on where (*i.e.*, locally, Sanger Imputation Server, Michigan Imputation Server, etc.) and with which softwares (*i.e.*, impute2, PBWT, etc.) you used to impute your genetic data, this information might be stored as `INFO` or `R2` (*most frequent names used, but it might be something else*).

#### Set the parameters

A template script (*i.e.*, `~/WP3/scripts/utils/RunAnalysis.R`) is available to help run the analysis as a bash/shell command.

1. Set the output directory.  
    The default outputs will be generated in the current directory (`./`).
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[3]}
    ```
2. Set a cohort name.  
    It's a identifier for a cohort and array.
    For example, "DESIR_Metabochip" for the results from the Metabochip Array in the DESIR cohort.
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[6]}
    ```
3. Set the number of CPUs you are going to use to run the Linear Mixed Model.
    Just to know who performed the analysis, if there are questions at a later point.
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[8]}
    ```
4. Set the path to your file `opal_credentials.txt`.
    For example, `~/WP3/scripts/opal_credentials.txt`, if you modified the default file (it is better to use absolute path, *i.e.*, without `~/`).
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[10]}
    ```
5. Set the path to your (RAW) VCF files
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[12]}
    ```
6. Set the imputation quality tag for your VCF
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[13]}
    ```
7. Set the path to VCFtools.  
    Default location of the software is usually `/usr/local/bin`.
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[14]}
    ```
8. Define if you want to format/split the VCF files.  
    Default is to format and split the VCF files available within the script.
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[16]}
    ```
9. Define to which part of the analysis you want to go.  
    1. Node access
    2. Phenotype QC
    3. Run Mixed Model without variants
    4. Check VCFs files
    5. Format VCFs files
    6. Run Mixed Model with (imputed) variants
    7. Check output files 
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[17]}
    ```
10. Define if genetic variants analyses should be performed.  
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[18]}
    ```
11. Define if genetic components should be used.  
    ```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")[20]}
    ```

Template: `~/WP3/scripts/utils/RunAnalysis.R`

```{r, eval = FALSE, code = readLines("~/WP3/scripts/utils/RunAnalysis.R")}
```

#### Start the analysis

##### Standard use

Once you created a R script using the template (*i.e.*, `~/WP3/scripts/utils/RunAnalysis.R`), you can start the analysis with a simple command.

```{sh, eval = FALSE}
Rscript ~/WP3/scripts/utils/RunAnalysis.R
```

##### Non-standard use

In the case you want to format your VCF files before running the script.


1. The best way to proceed is to use/set the parameters as showed below.
    ```{r, eval = FALSE}
    format_vcfs <- FALSE
    analysis_step <- 4
    variants_analysis <- FALSE
    ```

2. Run the script within bash/shell
    ```{sh, eval = FALSE}
    Rscript ~/WP3/scripts/utils/RunAnalysis.R
    ```
    This will create (*i.e.*, in the output directory you set up):
    * `./RHAPSODY_WP3_PreDiab_Cohort_Name_step4.html`  
        The report with the phenotype QC and without the genetic variant analyses.
    * `./vcffilespath.txt`  
        The list of all available VCF files to format using `./handleVCF_Cohort_Name.sh`.
        The files are also listed in the report in *Section 6.2.1 Check available VCF files*.
        ```{sh, eval = FALSE}
        /media/vcf/chr1.vcf.gz
        /media/vcf/chr2.vcf.gz
        ...
        /media/vcf/chr*.vcf.gz
        ...
        /media/vcf/chr22.vcf.gz
        ```
    * `./handleVCF_Cohort_Name.sh`  
        The command (with the correct parameters) to format the VCF using `~/WP3/scripts/utils/handleVCF.sh`.
        This is also stated in the report in *Section 6.2.2.3.1 Manually format VCFs*.
        ```{sh, eval = FALSE}
        #!/bin/sh
        
        ./handleVCF.sh Cohort_Name /media/vcf INFO /usr/local/bin 
        ```

3. Run the script to format the VCF
    ```{sh, eval = FALSE}
    sh ./handleVCF_Cohort_Name.sh
    ```

4. Run the whole analysis with the proper parameter
    ```{r, eval = FALSE}
    format_vcfs <- FALSE # keep this to FALSE
    analysis_step <- 7
    variants_analysis <- TRUE
    ```
    ```{sh, eval = FALSE}
    Rscript ~/WP3/scripts/utils/RunAnalysis.R
    ```


## Using Docker & Rstudio-server

### Install Docker

If Docker is not installed on your cluster/grid/computer/laptop, you can install it following the instructions on [docs.docker.com](https://docs.docker.com/).

### Deploy an image within a container

#### Pull the image

You can download the Docker image from Docker Hub: [/r/umr1283/rhapsody](https://hub.docker.com/r/umr1283/rhapsody).

```{sh, eval = FALSE}
docker pull umr1283/rhapsody
```

To pull a particular version (*i.e.*, not the latest), you can add the tag version.  
Tags are listed in the repository [/r/umr1283/rhapsody/tags](https://hub.docker.com/r/umr1283/rhapsody/tags).

```{sh, eval = FALSE}
docker pull umr1283/rhapsody:latest
```

#### Run a container

The easiest way to start a the Docker container using `umr1283/rhapsody` image is to run the following Docker command.

```{sh, eval = FALSE}
docker run \
  --name rhapsody \
  --detach \
  umr1283/rhapsody:latest
```

With those settings, the Docker container will not have access to any data stored in your cluster/grid/computer/laptop.  
In order to allow the Docker container to see some data, you have to use `--volume from:to` argument.

```{sh, eval = FALSE}
docker run \
  --name rhapsody \
  --detach \
  --volume /path/to/vcf:/media/vcf \
  umr1283/rhapsody:latest
```

If you want to more carefully set the Docker container, (, CPUs, memory, etc.), you can find all the parameters at the following address: [https://docs.docker.com/engine/reference/commandline/run/](https://docs.docker.com/engine/reference/commandline/run/)

### Connect to the container 

Open a web browser and type [http://localhost:8787](http://localhost:8787).  
You will see an authentication box with `Username` and `Password`:
    * Username: `rhapsody`
    * Password: `wp3`

### Files & Scripts

#### Retrieve the latest version

In order to retrieve the latest version of the main script `RHAPSODY_WP3_PreDiab.Rmd` and `utils/*` scripts, you can copy/paste the git commands below.  
This will pull (*i.e.*, download) the scripts from **GitHub**.  

```{sh, eval = FALSE}
sudo git -C /home/rhapsody/WP3/scripts pull origin master
```

Once donwload is complete, you should get the following directory tree in your home (`` `cd ~` ``).

```{r, echo = FALSE, results = "markup"}
outout_tree_structure <- c(
  list.files(
    path = "/home/rhapsody/WP3/scripts", 
    full.names = TRUE, include.dirs = TRUE
  ), 
  list.files(
    path = "/home/rhapsody/WP3/scripts", 
    full.names = TRUE, include.dirs = TRUE
  ) %>%
    list.files(full.names = TRUE)
) %>%
  unlist() %>%
  data.frame(pathString = .) %>%
  as.Node()

cat(paste(sapply(capture.output(outout_tree_structure)[-1], paste0, "\n"), collapse = ""))
```

#### Description

* `/home/rhapsody/WP3/scripts/README.pdf`  
  a short README file;
* `/home/rhapsody/WP3/scripts/RHAPSODY_WP3_PreDiab.Rmd`  
  a Rmarkdown script which perform:
    * opal node access,
    * phenotype QC, 
    * preliminary modelling,
    * VCFs formatting/filtering,
    * variants analysis;
* `/home/rhapsody/WP3/scripts/opal_credentials.txt`  
  the server, login and password informations to access phenotype data on your local node;
* `/home/rhapsody/WP3/scripts/utils/RHAPSODY_WP3_PreDiab_DEBUG.R`,  
  a small part of the main Rmarkdown script, which includes only node access and phenotype QC;
* `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`  
  a example of a R script to run the whole analysis in bash (parameters have to be filled properly according to your informations)  
  using `` `Rscript ~/WP3/scripts/utils/RunAnalysis.R` ``;
* `/home/rhapsody/WP3/scripts/utils/handleVCF.sh`  
  the bash/shell script used to format/filter VCFs files prior to the variant analysis.  
  This bash/shell script is also included within the main script, but you can use it directly as a standalone, if so (or if your already done formatting and don't want to do it again), parameter `format_vcfs` must be set to `FALSE`).
* `/home/rhapsody/WP3/scripts/utils/Install_Rpackages.R`  
  a R script which install a predefined version of all R packages loaded in the Rmarkdown script, only if needed (also avaiable within the main script);
* `/home/rhapsody/WP3/scripts/utils/check_analysis.Rmd`  
  a Rmarkdown script to ''quickly'' check if the analyses went well (*i.e.*, compute quantile-quatile plot, manhattan plot and a plot to compare chromosomal position to sequencial position);
* `/home/rhapsody/WP3/scripts/utils/README.Rmd`  
  a Rmarkdown script to create `README.pdf`.
* `/home/rhapsody/WP3/scripts/utils/howto.Rmd`  
  a how-to document and its html version `howto.html`.

### Run the analysis

#### OPAL database credentials

To avoid OPAL database credentials to be hard written within the script, you should create a text file with the `server`, `username` and `password` to access your local OPAL database (*i.e.*, RHAPSODY node hosting the phenotype data).  
You can also modify the default file `/home/rhapsody/WP3/scripts/opal_credentials.txt`.

```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/opal_credentials.txt")}
```

**NOTE:** The `username` provided in that file should have **"download"** rights.  
The data are downloaded locally in the R session for and during the analysis, then all local data are deleted (*i.e.*, the data are not leaving the R session in any way).  

#### Imputation quality (VCF)

Please check where the imputation quality is stored in your VCF files.  
Depending on where (*i.e.*, locally, Sanger Imputation Server, Michigan Imputation Server, etc.) and with which softwares (*i.e.*, impute2, PBWT, etc.) you used to impute your genetic data, this information might be stored as `INFO` or `R2` (*most frequent names used, but it might be something else*).

#### Set the parameters

A template script (*i.e.*, `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`) is available to help run the analysis as a bash/shell command.

1. Set the output directory.  
    The default outputs will be generated in the current directory (`./`).
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[3]}
    ```
2. Set a cohort name.  
    It's a identifier for a cohort and array.
    For example, "DESIR_Metabochip" for the results from the Metabochip Array in the DESIR cohort.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[6]}
    ```
3. Set the number of CPUs you are going to use to run the Linear Mixed Model.
    Just to know who performed the analysis, if there are questions at a later point.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[8]}
    ```
4. Set the path to your file `opal_credentials.txt`.
    For example, `~/WP3/scripts/opal_credentials.txt`, if you modified the default file (it is better to use absolute path, *i.e.*, without `~/`).
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[10]}
    ```
5. Set the path to your (RAW) VCF files
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[12]}
    ```
6. Set the imputation quality tag for your VCF
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[13]}
    ```
7. Set the path to VCFtools.  
    Default location of the software is usually `/usr/local/bin`.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[14]}
    ```
8. Define if you want to format/split the VCF files.  
    Default is to format and split the VCF files available within the script.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[16]}
    ```
9. Define to which part of the analysis you want to go.  
    1. Node access
    2. Phenotype QC
    3. Run Mixed Model without variants
    4. Check VCFs files
    5. Format VCFs files
    6. Run Mixed Model with (imputed) variants
    7. Check output files 
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[17]}
    ```
10. Define if genetic variants analyses should be performed.  
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[18]}
    ```
11. Define if genetic components should be used.  
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[20]}
    ```

Template: `~/WP3/scripts/utils/RunAnalysis.R`

```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")}
```

#### Start the analysis

##### Standard use

Once you created a R script using the template (*i.e.*, `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`), you can start the analysis with a simple command.

```{sh, eval = FALSE}
Rscript /home/rhapsody/WP3/scripts/utils/RunAnalysis.R
```

##### Non-standard use

In the case you want to format your VCF files before running the script.


1. The best way to proceed is to use/set the parameters as showed below.
    ```{r, eval = FALSE}
    format_vcfs <- FALSE
    analysis_step <- 4
    variants_analysis <- FALSE
    ```

2. Run the script within bash/shell
    ```{sh, eval = FALSE}
    Rscript /home/rhapsody/WP3/scripts/utils/RunAnalysis.R
    ```
    This will create (*i.e.*, in the output directory you set up):
    * `./RHAPSODY_WP3_PreDiab_Cohort_Name_step4.html`  
        The report with the phenotype QC and without the genetic variant analyses.
    * `./vcffilespath.txt`  
        The list of all available VCF files to format using `./handleVCF_Cohort_Name.sh`.
        The files are also listed in the report in *Section 6.2.1 Check available VCF files*.
        ```{sh, eval = FALSE}
        /media/vcf/chr1.vcf.gz
        /media/vcf/chr2.vcf.gz
        ...
        /media/vcf/chr*.vcf.gz
        ...
        /media/vcf/chr22.vcf.gz
        ```
    * `./handleVCF_Cohort_Name.sh`  
        The command (with the correct parameters) to format the VCF using `/home/rhapsody/WP3/scripts/utils/handleVCF.sh`.
        This is also stated in the report in *Section 6.2.2.3.1 Manually format VCFs*.
        ```{sh, eval = FALSE}
        #!/bin/sh
        
        ./handleVCF.sh Cohort_Name /media/vcf INFO /usr/local/bin 
        ```

3. Run the script to format the VCF
    ```{sh, eval = FALSE}
    sh ./handleVCF_Cohort_Name.sh
    ```

4. Run the whole analysis with the proper parameter
    ```{r, eval = FALSE}
    format_vcfs <- FALSE # keep this to FALSE
    analysis_step <- 7
    variants_analysis <- TRUE
    ```
    ```{sh, eval = FALSE}
    Rscript /home/rhapsody/WP3/scripts/utils/RunAnalysis.R
    ```


## Using Docker as an app

### Install Docker

If Docker is not installed on your cluster/grid/computer/laptop, you can install it following the instructions on [docs.docker.com](https://docs.docker.com/).

### Files & Scripts

#### Retrieve the latest version

In order to retrieve the latest version of the main script `RHAPSODY_WP3_PreDiab.Rmd` and `utils/*` scripts, you can copy/paste the git commands below.  
This will pull (*i.e.*, download) the scripts from **GitHub**.  

```{sh, eval = FALSE}
sudo git -C /home/rhapsody/WP3/scripts pull origin master
```

Once donwload is complete, you should get the following directory tree in your home (`` `cd ~` ``).

```{r, echo = FALSE, results = "markup"}
outout_tree_structure <- c(
  list.files(
    path = "/home/rhapsody/WP3/scripts", 
    full.names = TRUE, include.dirs = TRUE
  ), 
  list.files(
    path = "/home/rhapsody/WP3/scripts", 
    full.names = TRUE, include.dirs = TRUE
  ) %>%
    list.files(full.names = TRUE)
) %>%
  unlist() %>%
  data.frame(pathString = .) %>%
  as.Node()

cat(paste(sapply(capture.output(outout_tree_structure)[-1], paste0, "\n"), collapse = ""))
```

#### Description

* `/home/rhapsody/WP3/scripts/README.pdf`  
  a short README file;
* `/home/rhapsody/WP3/scripts/RHAPSODY_WP3_PreDiab.Rmd`  
  a Rmarkdown script which perform:
    * opal node access,
    * phenotype QC, 
    * preliminary modelling,
    * VCFs formatting/filtering,
    * variants analysis;
* `/home/rhapsody/WP3/scripts/opal_credentials.txt`  
  the server, login and password informations to access phenotype data on your local node;
* `/home/rhapsody/WP3/scripts/utils/RHAPSODY_WP3_PreDiab_DEBUG.R`,  
  a small part of the main Rmarkdown script, which includes only node access and phenotype QC;
* `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`  
  a example of a R script to run the whole analysis in bash (parameters have to be filled properly according to your informations)  
  using `` `Rscript ~/WP3/scripts/utils/RunAnalysis.R` ``;
* `/home/rhapsody/WP3/scripts/utils/handleVCF.sh`  
  the bash/shell script used to format/filter VCFs files prior to the variant analysis.  
  This bash/shell script is also included within the main script, but you can use it directly as a standalone, if so (or if your already done formatting and don't want to do it again), parameter `format_vcfs` must be set to `FALSE`).
* `/home/rhapsody/WP3/scripts/utils/Install_Rpackages.R`  
  a R script which install a predefined version of all R packages loaded in the Rmarkdown script, only if needed (also avaiable within the main script);
* `/home/rhapsody/WP3/scripts/utils/check_analysis.Rmd`  
  a Rmarkdown script to ''quickly'' check if the analyses went well (*i.e.*, compute quantile-quatile plot, manhattan plot and a plot to compare chromosomal position to sequencial position);
* `/home/rhapsody/WP3/scripts/utils/README.Rmd`  
  a Rmarkdown script to create `README.pdf`.
* `/home/rhapsody/WP3/scripts/utils/howto.Rmd`  
  a how-to document and its html version `howto.html`.

### Run the analysis

#### OPAL database credentials

To avoid OPAL database credentials to be hard written within the script, you should create a text file with the `server`, `username` and `password` to access your local OPAL database (*i.e.*, RHAPSODY node hosting the phenotype data).  
You can also modify the default file `/home/rhapsody/WP3/scripts/opal_credentials.txt`.

```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/opal_credentials.txt")}
```

**NOTE:** The `username` provided in that file should have **"download"** rights.  
The data are downloaded locally in the R session for and during the analysis, then all local data are deleted (*i.e.*, the data are not leaving the R session in any way).  

#### Imputation quality (VCF)

Please check where the imputation quality is stored in your VCF files.  
Depending on where (*i.e.*, locally, Sanger Imputation Server, Michigan Imputation Server, etc.) and with which softwares (*i.e.*, impute2, PBWT, etc.) you used to impute your genetic data, this information might be stored as `INFO` or `R2` (*most frequent names used, but it might be something else*).

#### Set the parameters

A template script (*i.e.*, `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`) is available to help run the analysis as a bash/shell command.

1. Set the output directory.  
    The default outputs will be generated in the current directory (`./`).
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[3]}
    ```
2. Set a cohort name.  
    It's a identifier for a cohort and array.
    For example, "DESIR_Metabochip" for the results from the Metabochip Array in the DESIR cohort.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[6]}
    ```
3. Set the number of CPUs you are going to use to run the Linear Mixed Model.
    Just to know who performed the analysis, if there are questions at a later point.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[8]}
    ```
4. Set the path to your file `opal_credentials.txt`.
    For example, `/home/rhapsody/WP3/scripts/opal_credentials.txt`, if you modified the default file (it is better to use absolute path, *i.e.*, without `~/`).
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[10]}
    ```
5. Set the path to your (RAW) VCF files
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[12]}
    ```
6. Set the imputation quality tag for your VCF
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[13]}
    ```
7. Set the path to VCFtools.  
    Default location of the software is usually `/usr/local/bin`.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[14]}
    ```
8. Define if you want to format/split the VCF files.  
    Default is to format and split the VCF files available within the script.
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[16]}
    ```
9. Define to which part of the analysis you want to go.  
    1. Node access
    2. Phenotype QC
    3. Run Mixed Model without variants
    4. Check VCFs files
    5. Format VCFs files
    6. Run Mixed Model with (imputed) variants
    7. Check output files 
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[17]}
    ```
10. Define if genetic variants analyses should be performed.  
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[18]}
    ```
11. Define if genetic components should be used.  
    ```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")[20]}
    ```

Template: `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`

```{r, eval = FALSE, code = readLines("/home/rhapsody/WP3/scripts/utils/RunAnalysis.R")}
```

#### Start the analysis

##### Standard use

Once you created a R script using the template (*i.e.*, `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R`), you can start the analysis with a simple command.

```{sh, eval = FALSE}
docker run \
  --name rhapsody \
  --detach \
  --volume /path/to/vcf/:/media/vcf \
  --volume /path/to/opal_credentials:/media/credentials \
  --volume /path/to/RunAnalysis:/media/RunAnalysis \
  --volume /path/to/rhapsody_output:/media/rhapsody_output \
  --rm \
  umr1283/rhapsody:latest Rscript /media/RunAnalysis/RunAnalysis.R
```

In that configuration, the script `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R` will be as follow:

```{r, eval = FALSE}
# set the output directory or leave as is; 
# output will be generated where the Rmarkdown file is
working_directory <- "/media/rhapsody_output" 

cohort_name <- "Cohort_Name"
author_name <- "Firstname LASTNAME"

n_cpu <- 2

opal_credentials <- "/media/credentials/opal_credentials.txt"

vcf_directory <- "/media/vcf"
imputation_quality_tag <- "INFO" # To be set according to VCF (could also be "R2")
vcftools_binary_path <- "/usr/local/bin"

format_vcfs <- TRUE
analysis_step <- 7
variants_analysis <- TRUE

genomic_component <- NULL # A csv file with SUBJID as first column and PC01 to PCXX



# Run the analysis
dir.create(path = working_directory, showWarnings = FALSE, mode = "0777")
rmarkdown::render(
  input = "/home/rhapsody/WP3/scripts/RHAPSODY_WP3_PreDiab.Rmd", 
  output_format = "html_document", 
  output_file = paste0(
    "RHAPSODY_WP3_PreDiab_", 
    cohort_name, "_step", 
    analysis_step, ".html"
  ), 
  output_dir = working_directory,
  params = list(
    cohort_name = cohort_name, 
    author_name = author_name, 
    opal_credentials = opal_credentials,
    vcf_input_directory = vcf_directory, 
    imputation_quality_tag = imputation_quality_tag, 
    vcftools_binary_path = vcftools_binary_path,
    output_directory = working_directory,
    analysis_step = analysis_step,
    format_vcfs = format_vcfs,
    variants_analysis = variants_analysis,
    chunk_size = 1000,
    exclude_X = TRUE,
    genomic_component = genomic_component,
    n_cpu = n_cpu,
    echo = FALSE, # Should R code be printed in the report
    warning = FALSE, # Should warnings be printed in the report
    message = FALSE, # Should messages be printed in the report
    debug = FALSE
  ),
  encoding = "UTF-8"
)
```

##### Non-standard use

In the case you want to format your VCF files before running the script.


1. The best way to proceed is to use/set the parameters as showed below.
    ```{r, eval = FALSE}
    format_vcfs <- FALSE
    analysis_step <- 4
    variants_analysis <- FALSE
    ```

2. Run the script within bash/shell
    ```{sh, eval = FALSE}
    docker run \
      --name rhapsody \
      --detach \
      --volume /path/to/vcf/:/media/vcf \
      --volume /path/to/opal_credentials:/media/credentials \
      --volume /path/to/RunAnalysis:/media/RunAnalysis \
      --volume /path/to/rhapsody_output:/media/rhapsody_output \
      --rm \
      umr1283/rhapsody:latest Rscript /media/RunAnalysis/RunAnalysis.R
    ```
    In that configuration, the script `/home/rhapsody/WP3/scripts/utils/RunAnalysis.R` will be as follow:
    ```{r, eval = FALSE}
    # set the output directory or leave as is; 
    # output will be generated where the Rmarkdown file is
    working_directory <- "/media/rhapsody_output" 
    
    cohort_name <- "Cohort_Name"
    author_name <- "Firstname LASTNAME"
    
    n_cpu <- 2
    
    opal_credentials <- "/media/credentials/opal_credentials.txt"
    
    vcf_directory <- "/media/vcf"
    imputation_quality_tag <- "INFO" # To be set according to VCF (could also be "R2")
    vcftools_binary_path <- "/usr/local/bin"
    
    format_vcfs <- FALSE
    analysis_step <- 4
    variants_analysis <- FALSE
    
    genomic_component <- NULL # A csv file with SUBJID as first column and PC01 to PCXX
    
    
    
    # Run the analysis
    dir.create(path = working_directory, showWarnings = FALSE, mode = "0777")
    rmarkdown::render(
      input = "/home/rhapsody/WP3/scripts/RHAPSODY_WP3_PreDiab.Rmd", 
      output_format = "html_document", 
      output_file = paste0(
        "RHAPSODY_WP3_PreDiab_", 
        cohort_name, "_step", 
        analysis_step, ".html"
      ), 
      output_dir = working_directory,
      params = list(
        cohort_name = cohort_name, 
        author_name = author_name, 
        opal_credentials = opal_credentials,
        vcf_input_directory = vcf_directory, 
        imputation_quality_tag = imputation_quality_tag, 
        vcftools_binary_path = vcftools_binary_path,
        output_directory = working_directory,
        analysis_step = analysis_step,
        format_vcfs = format_vcfs,
        variants_analysis = variants_analysis,
        chunk_size = 1000,
        exclude_X = TRUE,
        genomic_component = genomic_component,
        n_cpu = n_cpu,
        echo = FALSE, # Should R code be printed in the report
        warning = FALSE, # Should warnings be printed in the report
        message = FALSE, # Should messages be printed in the report
        debug = FALSE
      ),
      encoding = "UTF-8"
    )

    ```
    This will create (*i.e.*, in the output directory you set up):
    * `/media/rhapsody_output/RHAPSODY_WP3_PreDiab_Cohort_Name_step4.html`  
        The report with the phenotype QC and without the genetic variant analyses.
    * `/media/rhapsody_output/vcffilespath.txt`  
        The list of all available VCF files to format using `./handleVCF_Cohort_Name.sh`.
        The files are also listed in the report in *Section 6.2.1 Check available VCF files*.
        ```{sh, eval = FALSE}
        /media/vcf/chr1.vcf.gz
        /media/vcf/chr2.vcf.gz
        ...
        /media/vcf/chr*.vcf.gz
        ...
        /media/vcf/chr22.vcf.gz
        ```
    * `/media/rhapsody_output/handleVCF_Cohort_Name.sh`  
        The command (with the correct parameters) to format the VCF using `~/WP3/scripts/utils/handleVCF.sh`.
        This is also stated in the report in *Section 6.2.2.3.1 Manually format VCFs*.
        ```{sh, eval = FALSE}
        #!/bin/sh
        
        ./handleVCF.sh Cohort_Name /media/vcf INFO /usr/local/bin 
        ```

3. Run the script to format the VCF
    ```{sh, eval = FALSE}
    docker run \
      --name rhapsody \
      --detach \
      --volume /path/to/vcf/:/media/vcf \
      --volume /path/to/opal_credentials:/media/credentials \
      --volume /path/to/RunAnalysis:/media/RunAnalysis \
      --volume /path/to/rhapsody_output:/media/rhapsody_output \
      --rm \
      umr1283/rhapsody:latest sh /media/rhapsody_output/handleVCF_Cohort_Name.sh
    ```
    
4. Run the whole analysis with the proper parameter
    ```{r, eval = FALSE}
    format_vcfs <- FALSE # keep this to FALSE
    analysis_step <- 7
    variants_analysis <- TRUE
    ```
    ```{sh, eval = FALSE}
    docker run \
      --name rhapsody \
      --detach \
      --volume /path/to/vcf/:/media/vcf \
      --volume /path/to/opal_credentials:/media/credentials \
      --volume /path/to/RunAnalysis:/media/RunAnalysis \
      --volume /path/to/rhapsody_output:/media/rhapsody_output \
      --rm \
      umr1283/rhapsody:latest Rscript /media/RunAnalysis/RunAnalysis.R
    ```
    In that configuration, the script `~/WP3/scripts/utils/RunAnalysis.R` will be as follow:
    ```{r, eval = FALSE}
    # set the output directory or leave as is; 
    # output will be generated where the Rmarkdown file is
    working_directory <- "/media/rhapsody_output" 
    
    cohort_name <- "Cohort_Name"
    author_name <- "Firstname LASTNAME"
    
    n_cpu <- 2
    
    opal_credentials <- "/media/credentials/opal_credentials.txt"
    
    vcf_directory <- "/media/vcf"
    imputation_quality_tag <- "INFO" # To be set according to VCF (could also be "R2")
    vcftools_binary_path <- "/usr/local/bin"
    
    format_vcfs <- FALSE # keep this to FALSE
    analysis_step <- 7
    variants_analysis <- TRUE
    
    genomic_component <- NULL # A csv file with SUBJID as first column and PC01 to PCXX
    
    
    
    # Run the analysis
    dir.create(path = working_directory, showWarnings = FALSE, mode = "0777")
    rmarkdown::render(
      input = "/home/rhapsody/WP3/scripts/RHAPSODY_WP3_PreDiab.Rmd", 
      output_format = "html_document", 
      output_file = paste0(
        "RHAPSODY_WP3_PreDiab_", 
        cohort_name, "_step", 
        analysis_step, ".html"
      ), 
      output_dir = working_directory,
      params = list(
        cohort_name = cohort_name, 
        author_name = author_name, 
        opal_credentials = opal_credentials,
        vcf_input_directory = vcf_directory, 
        imputation_quality_tag = imputation_quality_tag, 
        vcftools_binary_path = vcftools_binary_path,
        output_directory = working_directory,
        analysis_step = analysis_step,
        format_vcfs = format_vcfs,
        variants_analysis = variants_analysis,
        chunk_size = 1000,
        exclude_X = TRUE,
        genomic_component = genomic_component,
        n_cpu = n_cpu,
        echo = FALSE, # Should R code be printed in the report
        warning = FALSE, # Should warnings be printed in the report
        message = FALSE, # Should messages be printed in the report
        debug = FALSE
      ),
      encoding = "UTF-8"
    )
    ```
